<!DOCTYPE html>



  


<html class="theme-next muse use-motion" lang="">
<head>
  <meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>
<meta name="theme-color" content="#222">



  
  
    
    
  <script src="/lib/pace/pace.min.js?v=1.0.2"></script>
  <link href="/lib/pace/pace-theme-minimal.min.css?v=1.0.2" rel="stylesheet">







<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />



  <meta name="google-site-verification" content="google5b248f7b86cbcee5.html" />














  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.4" rel="stylesheet" type="text/css" />


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.4">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon.ico?v=5.1.4">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.4">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.4" color="#222">





  <meta name="keywords" content="assignment,techniques,bioinformatics," />





  <link rel="alternate" href="/atom.xml" title="WonderLand" type="application/atom+xml" />






<meta name="description" content="It contains my assignment in Bioinformatics course talking about Differential Expression and this semesters’ code">
<meta name="keywords" content="assignment,techniques,bioinformatics">
<meta property="og:type" content="article">
<meta property="og:title" content="Differential Expression&amp;作业">
<meta property="og:url" content="https://www.cmwonderland.com/2018/05/20/diffexpress/index.html">
<meta property="og:site_name" content="WonderLand">
<meta property="og:description" content="It contains my assignment in Bioinformatics course talking about Differential Expression and this semesters’ code">
<meta property="og:locale" content="default">
<meta property="og:updated_time" content="2018-05-21T13:08:07.966Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Differential Expression&amp;作业">
<meta name="twitter:description" content="It contains my assignment in Bioinformatics course talking about Differential Expression and this semesters’ code">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Muse',
    version: '5.1.4',
    sidebar: {"position":"left","display":"always","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: 'Author'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="https://www.cmwonderland.com/2018/05/20/diffexpress/"/>





  <title>Differential Expression&作业 | WonderLand</title>
  





  <script type="text/javascript">
    var _hmt = _hmt || [];
    (function() {
      var hm = document.createElement("script");
      hm.src = "https://hm.baidu.com/hm.js?6656499bfc0e07b4e20ee4975eb85f31";
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(hm, s);
    })();
  </script>




</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="default">

  
  
    
  

  <div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <a href="https://github.com/james20141606"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_white_ffffff.png" alt="Fork me on GitHub"></a>
    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">WonderLand</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle">Somnium & Somniator</p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            Home
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            About
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            Tags
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            Categories
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            Archives
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="https://www.cmwonderland.com/2018/05/20/diffexpress/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="James Chen">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.png">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="WonderLand">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">Differential Expression&作业</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              
              <time title="Post created" itemprop="dateCreated datePublished" datetime="2018-05-20T00:08:12-04:00">
                2018-05-20
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/techniques/" itemprop="url" rel="index">
                    <span itemprop="name">techniques</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          
             <span id="/2018/05/20/diffexpress/" class="leancloud_visitors" data-flag-title="Differential Expression&作业">
               <span class="post-meta-divider">|</span>
               <span class="post-meta-item-icon">
                 <i class="fa fa-eye"></i>
               </span>
               
                 <span class="post-meta-item-text">Visitors&#58;</span>
               
                 <span class="leancloud-visitors-count"></span>
             </span>
          

          

          
            <div class="post-wordcount">
              
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article&#58;</span>
                
                <span title="Words count in article">
                  1,679
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time &asymp;</span>
                
                <span title="Reading time">
                  10
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <p>It contains my assignment in Bioinformatics course talking about Differential Expression and this semesters’ code<br><a id="more"></a></p>
<h1 id="My-assignment-in-bioinformatics-course"><a href="#My-assignment-in-bioinformatics-course" class="headerlink" title="My assignment in bioinformatics course"></a>My assignment in bioinformatics course</h1>

	<div class="row">
		<iframe src="https://drive.google.com/file/d/1d2YKSKrCfnELAHiAlEqAxIBErdCfR4-L/preview" style="width:100%; height:550px"></iframe>
	</div>



<h1 id="Codes-using-different-packages-producing-differential-expression-files"><a href="#Codes-using-different-packages-producing-differential-expression-files" class="headerlink" title="Codes using different packages producing differential expression files"></a>Codes using different packages producing differential expression files</h1><figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br></pre></td><td class="code"><pre><span class="line">setwd(<span class="string">'61samplesmatrix'</span>)</span><br><span class="line">#rnanames = <span class="string">'miRNA'</span></span><br><span class="line">rnaname = <span class="string">'miRNA'</span></span><br><span class="line">commonpath = <span class="string">'04.counts/hcc_lulab.sequentialMap.homer.'</span></span><br><span class="line">savepath = <span class="string">'05.diffexp/hcc_lulab.sequentialMap.homer.'</span></span><br><span class="line">mx &lt;- read.table(paste(commonpath,rnaname,<span class="string">".mx"</span>,sep=<span class="string">""</span>),sep=<span class="string">"\t"</span>,header=T)</span><br><span class="line">design &lt;- read.table(<span class="string">"05.diffexp/design.tsv"</span>,sep=<span class="string">"\t"</span>,header=T)</span><br><span class="line">filter_genes &lt;- apply(</span><br><span class="line">  mx[,<span class="number">2</span>:ncol(mx)],</span><br><span class="line">  <span class="number">1</span>,</span><br><span class="line">  function(x) length(x[x &gt; <span class="number">2</span>]) &gt;= <span class="number">2</span></span><br><span class="line">)</span><br><span class="line">mx_filterGenes &lt;- mx[filter_genes,]</span><br><span class="line">####################################################################################</span><br><span class="line">####################################################################################</span><br><span class="line">library(DESeq2)</span><br><span class="line">countData &lt;- mx_filterGenes</span><br><span class="line">colData &lt;- design</span><br><span class="line">dds &lt;- DESeqDataSetFromMatrix(countData, colData, design=~Treatment, tidy=TRUE)</span><br><span class="line">norm &lt;- rlog(dds,blind=FALSE)</span><br><span class="line">norm_matrix &lt;- assay(norm)</span><br><span class="line">norm_df &lt;- data.frame(Gene=rownames(norm_matrix), norm_matrix)</span><br><span class="line"><span class="keyword">if</span>(!file.exists(paste(savepath,rnaname,<span class="string">".DESeq2.rlog.mx"</span>,sep=<span class="string">""</span>)))&#123;</span><br><span class="line">  write.table(norm_df, paste(savepath,rnaname,<span class="string">".DESeq2.rlog.mx"</span>,sep=<span class="string">""</span>), row.names = FALSE,sep=<span class="string">"\t"</span>)&#125;</span><br><span class="line">deg &lt;- DESeq(dds)</span><br><span class="line"></span><br><span class="line">res &lt;- results(deg,contrast=c(<span class="string">"Treatment"</span>,<span class="string">"normal"</span>,<span class="string">"Before"</span>),tidy=TRUE)</span><br><span class="line">merged_res &lt;- merge(norm_df,res,by.x=<span class="string">"Gene"</span>,by.y=<span class="string">"row"</span>)</span><br><span class="line"><span class="keyword">if</span>(!file.exists(paste(savepath,rnaname,<span class="string">".NormalvsBefore.DESeq2.tsv"</span>,sep=<span class="string">""</span>)))&#123;</span><br><span class="line">  write.table(merged_res,file=paste(savepath,rnaname,<span class="string">".NormalvsBefore.DESeq2.tsv"</span>,sep=<span class="string">""</span>),sep=<span class="string">"\t"</span>,row.names=FALSE)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">res1 &lt;- results(deg,contrast=c(<span class="string">"Treatment"</span>,<span class="string">"normal"</span>,<span class="string">"After"</span>),tidy=TRUE)</span><br><span class="line">merged_res1 &lt;- merge(norm_df,res1,by.x=<span class="string">"Gene"</span>,by.y=<span class="string">"row"</span>)</span><br><span class="line"><span class="keyword">if</span>(!file.exists(paste(savepath,rnaname,<span class="string">".NormalvsAfter.DESeq2.tsv"</span>,sep=<span class="string">""</span>)))&#123;</span><br><span class="line">  write.table(merged_res1,file=paste(savepath,rnaname,<span class="string">".NormalvsAfter.DESeq2.tsv"</span>,sep=<span class="string">""</span>),sep=<span class="string">"\t"</span>,row.names=FALSE)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">res2 &lt;- results(deg,contrast=c(<span class="string">"Treatment"</span>,<span class="string">"After"</span>,<span class="string">"Before"</span>),tidy=TRUE)</span><br><span class="line">merged_res2 &lt;- merge(norm_df,res2,by.x=<span class="string">"Gene"</span>,by.y=<span class="string">"row"</span>)</span><br><span class="line"><span class="keyword">if</span>(!file.exists(paste(savepath,rnaname,<span class="string">".AftervsBefore.DESeq2.tsv"</span>,sep=<span class="string">""</span>)))&#123;</span><br><span class="line">  write.table(merged_res2,file=paste(savepath,rnaname,<span class="string">".AftervsBefore.DESeq2.tsv"</span>,sep=<span class="string">""</span>),sep=<span class="string">"\t"</span>,row.names=FALSE)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">####################################################################################</span><br><span class="line">####################################################################################</span><br><span class="line">library(edgeR)</span><br><span class="line">#Read Data <span class="keyword">in</span></span><br><span class="line">countData &lt;- mx_filterGenes[,<span class="number">-1</span>]</span><br><span class="line">rownames(countData) &lt;- mx_filterGenes[,<span class="number">1</span>]</span><br><span class="line">design &lt;- read.table(<span class="string">"05.diffexp/design.tsv"</span>,sep=<span class="string">"\t"</span>,header=T)</span><br><span class="line">colData &lt;- design</span><br><span class="line">y &lt;- DGEList(countData, samples=colData, group=colData$Treatment)</span><br><span class="line">y &lt;- calcNormFactors(y)</span><br><span class="line">design &lt;-model.matrix(~Treatment, data=colData)</span><br><span class="line">y &lt;- estimateDisp(y, design)</span><br><span class="line">et &lt;- exactTest(y)</span><br><span class="line">res &lt;- topTags(et,Inf)</span><br><span class="line">tidyResult &lt;- data.frame(Gene=rownames(res$table), res$table)</span><br><span class="line"><span class="keyword">if</span>(!file.exists(paste(savepath,rnaname,<span class="string">".NCvsHCC.edgeR.classic.tsv"</span>,sep=<span class="string">""</span>)))&#123;</span><br><span class="line">  write.table(tidyResult,file=paste(savepath,rnaname,<span class="string">".NCvsHCC.edgeR.classic.tsv"</span>,sep=<span class="string">""</span>),sep=<span class="string">"\t"</span>,row.names=FALSE)&#125;</span><br><span class="line">fit &lt;- glmFit(y,design)</span><br><span class="line">lrt &lt;- glmLRT(fit,contrast = c(<span class="number">1</span>,<span class="number">-1</span>))</span><br><span class="line">FDR &lt;- p.adjust(lrt$table$PValue, method=<span class="string">"BH"</span>)</span><br><span class="line">padj_lrt &lt;- cbind(lrt$table,FDR)</span><br><span class="line">fit_df &lt;- lrt$fitted.values</span><br><span class="line"><span class="keyword">if</span>(!file.exists(paste(savepath,rnaname,<span class="string">".homer.edgeR.TMM.mx"</span>,sep=<span class="string">""</span>)))&#123;</span><br><span class="line">  write.table(fit_df,file = paste(savepath,rnaname,<span class="string">".homer.edgeR.TMM.mx"</span>,sep=<span class="string">""</span>),row.names = T, sep=<span class="string">"\t"</span>, quote=F)&#125;</span><br><span class="line">merged_lrt &lt;- merge(fit_df,padj_lrt,by=<span class="string">"row.names"</span>)</span><br><span class="line">colnames(merged_lrt)[<span class="number">1</span>] &lt;- <span class="string">"Genes"</span></span><br><span class="line"><span class="keyword">if</span>(!file.exists(paste(savepath,rnaname,<span class="string">".NCvsHCC.edgeR.tsv"</span>,sep=<span class="string">""</span>)))&#123;</span><br><span class="line">  write.table(merged_lrt,file =paste(savepath,rnaname,<span class="string">".NCvsHCC.edgeR.tsv"</span>,sep=<span class="string">""</span>),row.names = F, sep=<span class="string">"\t"</span>, quote=F)&#125;</span><br><span class="line">####################################################################################</span><br><span class="line">####################################################################################</span><br><span class="line"></span><br><span class="line">cpmMx &lt;- read.table(paste(commonpath,rnaname,<span class="string">".homer.rpm.m"</span>,sep=<span class="string">""</span>),sep=<span class="string">"\t"</span>,header=T)</span><br><span class="line">filter_cpm &lt;- apply(</span><br><span class="line">  mx[,<span class="number">2</span>:ncol(cpmMx)],</span><br><span class="line">  <span class="number">1</span>,</span><br><span class="line">  function(x) length(x[x &gt; <span class="number">0</span>]) &gt;= <span class="number">2</span></span><br><span class="line">)</span><br><span class="line">mx_filterCPM &lt;- cpmMx[filter_cpm,]</span><br><span class="line">myFun &lt;- function(x)&#123;</span><br><span class="line">  x = <span class="keyword">as</span>.numeric(x)</span><br><span class="line">  v1 = x[<span class="number">2</span>:<span class="number">4</span>]</span><br><span class="line">  v2 = x[<span class="number">5</span>:<span class="number">10</span>]</span><br><span class="line">  out &lt;- wilcox.test(v1,v2,exact = F) </span><br><span class="line">  out &lt;- out$p.value</span><br><span class="line">&#125;</span><br><span class="line">p_value &lt;- apply(mx_filterCPM,<span class="number">1</span>,myFun)</span><br><span class="line">p_value[is.nan(p_value)] &lt;- <span class="number">1</span></span><br><span class="line">FDR &lt;- p.adjust(p_value,method = <span class="string">"BH"</span>)</span><br><span class="line">mx_filterCPM$avgNC &lt;- apply(mx_filterCPM[,<span class="number">2</span>:<span class="number">4</span>],<span class="number">1</span>,mean)</span><br><span class="line">mx_filterCPM$avgHCC &lt;- apply(mx_filterCPM[,<span class="number">5</span>:<span class="number">10</span>],<span class="number">1</span>,mean)</span><br><span class="line">mx_filterCPM$log2fc &lt;- log2((mx_filterCPM$avgNC+<span class="number">0.25</span>)/(mx_filterCPM$avgHCC+<span class="number">0.25</span>))</span><br><span class="line">results &lt;- cbind(mx_filterCPM,p_value,FDR)</span><br><span class="line"><span class="keyword">if</span>(!file.exists(paste(savepath,rnaname,<span class="string">".NCvsHCC.wilcox.tsv"</span>,sep=<span class="string">""</span>)))&#123;</span><br><span class="line">  write.table(results,file = paste(savepath,rnaname,<span class="string">".NCvsHCC.wilcox.tsv"</span>,sep=<span class="string">""</span>),row.names = F, sep=<span class="string">"\t"</span>, quote=F)&#125;</span><br></pre></td></tr></table></figure>
<h1 id="some-plot-from-DESeq2"><a href="#some-plot-from-DESeq2" class="headerlink" title="some plot from DESeq2"></a>some plot from DESeq2</h1><p>Here is the pdf version of the following codes and its output.<br>

	<div class="row">
		<iframe src="https://drive.google.com/file/d/1TYMXYXPnZ_1SDT9FQgQUx8g0wy4QXGQm/preview" style="width:100%; height:550px"></iframe>
	</div>


</p>
<p>If you can’t see pdf, please try this more interactive way by <a href="https://www.cmwonderland.com/expression.html"><strong>click here</strong></a></p>
<figure class="highlight routeros"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">### DESeq2</span></span><br><span class="line">&#123;r <span class="attribute">warning</span>=<span class="literal">FALSE</span>&#125;</span><br><span class="line"><span class="comment"># filter expression matrix</span></span><br><span class="line">mx &lt;- read.table(<span class="string">"04.counts/hcc_example.miRNA.homer.ct.mx"</span>,<span class="attribute">sep</span>=<span class="string">"\t"</span>,header=T)</span><br><span class="line"></span><br><span class="line"><span class="comment"># filter genes</span></span><br><span class="line">filter_genes &lt;- apply(</span><br><span class="line">  mx[,2:ncol(mx)],</span><br><span class="line">  1,</span><br><span class="line">  function(x) length(x[x &gt; 2]) &gt;= 2</span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">mx_filterGenes &lt;- mx[filter_genes,]</span><br><span class="line"></span><br><span class="line"><span class="comment"># load experimential design</span></span><br><span class="line">design &lt;- read.table(<span class="string">"05.diffexp/design.tsv"</span>,<span class="attribute">sep</span>=<span class="string">"\t"</span>,header=T)</span><br><span class="line"></span><br><span class="line"><span class="comment">#-----------------------------------------</span></span><br><span class="line"><span class="comment"># basic script for normalizing with DESeq2</span></span><br><span class="line">library(DESeq2)</span><br><span class="line"><span class="comment">#Read Data in</span></span><br><span class="line">countData &lt;- mx_filterGenes</span><br><span class="line">colData &lt;- design</span><br><span class="line"></span><br><span class="line"><span class="comment"># generate Dataset</span></span><br><span class="line">dds &lt;- DESeqDataSetFromMatrix(countData, colData, <span class="attribute">design</span>=~Treatment, <span class="attribute">tidy</span>=<span class="literal">TRUE</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># normlize using rlog method</span></span><br><span class="line">norm &lt;- rlog(dds,<span class="attribute">blind</span>=<span class="literal">FALSE</span>)</span><br><span class="line">norm_matrix &lt;- assay(norm)</span><br><span class="line">norm_df &lt;- data.frame(<span class="attribute">Gene</span>=rownames(norm_matrix), norm_matrix)</span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span>(!file.exists(<span class="string">"05.diffexp/hcc_example.miRNA.homer.DESeq2.rlog.mx"</span>))&#123;</span><br><span class="line">write.table(norm_df, <span class="string">"05.diffexp/hcc_example.miRNA.homer.DESeq2.rlog.mx"</span>, row.names = <span class="literal">FALSE</span>,<span class="attribute">sep</span>=<span class="string">"\t"</span>)&#125;</span><br><span class="line"></span><br><span class="line">deg &lt;- DESeq(dds)</span><br><span class="line">res &lt;- results(deg,<span class="attribute">tidy</span>=<span class="literal">TRUE</span>)</span><br><span class="line">merged_res &lt;- merge(norm_df,res,by.<span class="attribute">x</span>=<span class="string">"Gene"</span>,by.y="row")</span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span>(!file.exists(<span class="string">"05.diffexp/hcc_example.miRNA.NCvsHCC.DESeq2.tsv"</span>))&#123;</span><br><span class="line">write.table(merged_res,<span class="attribute">file</span>=<span class="string">"05.diffexp/hcc_example.miRNA.NCvsHCC.DESeq2.tsv"</span>,sep="\t",row.names=FALSE)</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>the codes use DESeq2 and produces files:</p>
<p>05.diffexp/hcc_example.miRNA.homer.DESeq2.rlog.mx</p>
<p>05.diffexp/hcc_example.miRNA.NCvsHCC.DESeq2.tsv</p>
<h2 id="edgeR"><a href="#edgeR" class="headerlink" title="edgeR"></a>edgeR</h2><figure class="highlight plain"><figcaption><span>warning</span></figcaption><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br></pre></td><td class="code"><pre><span class="line"># 0.filter expression matrix</span><br><span class="line">mx &lt;- read.table(&quot;04.counts/hcc_example.miRNA.homer.ct.mx&quot;,sep=&quot;\t&quot;,header=T)</span><br><span class="line"></span><br><span class="line"># filter genes</span><br><span class="line">filter_genes &lt;- apply(</span><br><span class="line">    mx[,2:ncol(mx)],</span><br><span class="line">    1,</span><br><span class="line">    function(x) length(x[x &gt; 2]) &gt;= 2</span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">mx_filterGenes &lt;- mx[filter_genes,]</span><br><span class="line"></span><br><span class="line"># load experimential design</span><br><span class="line">design &lt;- read.table(&quot;05.diffexp/design.tsv&quot;,sep=&quot;\t&quot;,header=T)</span><br><span class="line"></span><br><span class="line">#--------------------------------</span><br><span class="line"># basic script for running edgeR</span><br><span class="line">library(edgeR)</span><br><span class="line">#Read Data in</span><br><span class="line">countData &lt;- mx_filterGenes[,-1]</span><br><span class="line">rownames(countData) &lt;- mx_filterGenes[,1]</span><br><span class="line">design &lt;- read.table(&quot;05.diffexp/design.tsv&quot;,sep=&quot;\t&quot;,header=T)</span><br><span class="line">colData &lt;- design</span><br><span class="line"></span><br><span class="line"># generate DGE object</span><br><span class="line">y &lt;- DGEList(countData, samples=colData, group=colData$Treatment)</span><br><span class="line">y &lt;- calcNormFactors(y)</span><br><span class="line"></span><br><span class="line">#Estimate Error Model</span><br><span class="line">design &lt;-model.matrix(~Treatment, data=colData)</span><br><span class="line">y &lt;- estimateDisp(y, design)</span><br><span class="line"></span><br><span class="line"># classic methods: compute p-values, then output</span><br><span class="line">et &lt;- exactTest(y)</span><br><span class="line">res &lt;- topTags(et,Inf)</span><br><span class="line">tidyResult &lt;- data.frame(Gene=rownames(res$table), res$table)</span><br><span class="line"></span><br><span class="line">if(!file.exists(&quot;05.diffexp/hcc_example.miRNA.NCvsHCC.edgeR.classic.tsv&quot;))&#123;</span><br><span class="line">write.table(tidyResult,file=&quot;05.diffexp/hcc_example.miRNA.NCvsHCC.edgeR.classic.tsv&quot;,sep=&quot;\t&quot;,row.names=FALSE)&#125;</span><br><span class="line"></span><br><span class="line"># Generalized linear models</span><br><span class="line">fit &lt;- glmFit(y,design)</span><br><span class="line"># likelihood ratio test</span><br><span class="line">lrt &lt;- glmLRT(fit,contrast = c(1,-1))</span><br><span class="line">FDR &lt;- p.adjust(lrt$table$PValue, method=&quot;BH&quot;)</span><br><span class="line">padj_lrt &lt;- cbind(lrt$table,FDR)</span><br><span class="line">fit_df &lt;- lrt$fitted.values</span><br><span class="line">if(!file.exists(&quot;05.diffexp/hcc_example.miRNA.homer.edgeR.TMM.mx&quot;))&#123;</span><br><span class="line">write.table(fit_df,file = &quot;05.diffexp/hcc_example.miRNA.homer.edgeR.TMM.mx&quot;,row.names = T, sep=&quot;\t&quot;, quote=F)&#125;</span><br><span class="line">merged_lrt &lt;- merge(fit_df,padj_lrt,by=&quot;row.names&quot;)</span><br><span class="line">colnames(merged_lrt)[1] &lt;- &quot;Genes&quot;</span><br><span class="line"></span><br><span class="line">if(!file.exists(&quot;05.diffexp/hcc_example.miRNA.NCvsHCC.edgeR.tsv&quot;))&#123;</span><br><span class="line">write.table(merged_lrt,file =&quot;05.diffexp/hcc_example.miRNA.NCvsHCC.edgeR.tsv&quot;,row.names = F, sep=&quot;\t&quot;, quote=F)&#125;</span><br></pre></td></tr></table></figure>
<p>the codes use edgeR and produces files:</p>
<p>05.diffexp/hcc_example.miRNA.NCvsHCC.edgeR.classic.tsv</p>
<p>05.diffexp/hcc_example.miRNA.homer.edgeR.TMM.mx</p>
<p>05.diffexp/hcc_example.miRNA.NCvsHCC.edgeR.tsv</p>
<h2 id="Wilcox-Mann-Whitney-U-Test"><a href="#Wilcox-Mann-Whitney-U-Test" class="headerlink" title="Wilcox/Mann-Whitney-U Test"></a>Wilcox/Mann-Whitney-U Test</h2><h3 id="1-normalize-the-reads-by-library-size-edgeR"><a href="#1-normalize-the-reads-by-library-size-edgeR" class="headerlink" title="1. normalize the reads by library size (edgeR)"></a>1. normalize the reads by library size (edgeR)</h3><h3 id="2-identify-differential-expressed-gene-using-wilcoxon-test"><a href="#2-identify-differential-expressed-gene-using-wilcoxon-test" class="headerlink" title="2. identify differential expressed gene using wilcoxon.test()"></a>2. identify differential expressed gene using wilcoxon.test()</h3><figure class="highlight plain"><figcaption><span>warning</span></figcaption><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line">cpmMx &lt;- read.table(&quot;04.counts/hcc_example.miRNA.homer.rpm.mx&quot;,sep=&quot;\t&quot;,header=T)</span><br><span class="line">filter_cpm &lt;- apply(</span><br><span class="line">    mx[,2:ncol(cpmMx)],</span><br><span class="line">    1,</span><br><span class="line">    function(x) length(x[x &gt; 0]) &gt;= 2</span><br><span class="line">)</span><br><span class="line">mx_filterCPM &lt;- cpmMx[filter_cpm,]</span><br><span class="line"></span><br><span class="line">myFun &lt;- function(x)&#123;</span><br><span class="line">  x = as.numeric(x)</span><br><span class="line">  v1 = x[2:4]</span><br><span class="line">  v2 = x[5:10]</span><br><span class="line">  out &lt;- wilcox.test(v1,v2,exact = F) </span><br><span class="line">  out &lt;- out$p.value</span><br><span class="line">&#125;</span><br><span class="line">p_value &lt;- apply(mx_filterCPM,1,myFun)</span><br><span class="line">p_value[is.nan(p_value)] &lt;- 1</span><br><span class="line">FDR &lt;- p.adjust(p_value,method = &quot;BH&quot;)</span><br><span class="line">mx_filterCPM$avgNC &lt;- apply(mx_filterCPM[,2:4],1,mean)</span><br><span class="line">mx_filterCPM$avgHCC &lt;- apply(mx_filterCPM[,5:10],1,mean)</span><br><span class="line">mx_filterCPM$log2fc &lt;- log2((mx_filterCPM$avgNC+0.25)/(mx_filterCPM$avgHCC+0.25))</span><br><span class="line">results &lt;- cbind(mx_filterCPM,p_value,FDR)</span><br><span class="line">if(!file.exists(&quot;05.diffexp/hcc_example.miRNA.NCvsHCC.wilcox__.tsv&quot;))&#123;</span><br><span class="line">write.table(results,file = &quot;05.diffexp/hcc_example.miRNA.NCvsHCC.wilcox__.tsv&quot;,row.names = F, sep=&quot;\t&quot;, quote=F)&#125;</span><br></pre></td></tr></table></figure>
<p>the codes use wilconx test and produces files:</p>
<p>05.diffexp/hcc<em>example.miRNA.NCvsHCC.wilcox</em>.tsv</p>
<h2 id="exploring-results"><a href="#exploring-results" class="headerlink" title="exploring results"></a>exploring results</h2><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line"># filter expression matrix</span><br><span class="line">mx &lt;- read.table(&quot;04.counts/hcc_example.miRNA.homer.ct.mx&quot;,sep=&quot;\t&quot;,header=T)</span><br><span class="line"></span><br><span class="line"># filter genes</span><br><span class="line">filter_genes &lt;- apply(</span><br><span class="line">  mx[,2:ncol(mx)],</span><br><span class="line">  1,</span><br><span class="line">  function(x) length(x[x &gt; 2]) &gt;= 2</span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">mx_filterGenes &lt;- mx[filter_genes,]</span><br><span class="line"></span><br><span class="line"># load experimential design</span><br><span class="line">design &lt;- read.table(&quot;05.diffexp/design.tsv&quot;,sep=&quot;\t&quot;,header=T)</span><br><span class="line"></span><br><span class="line">#-----------------------------------------</span><br><span class="line"># basic script for normalizing with DESeq2</span><br><span class="line">library(DESeq2)</span><br><span class="line">#Read Data in</span><br><span class="line">countData &lt;- mx_filterGenes</span><br><span class="line">colData &lt;- design</span><br><span class="line"></span><br><span class="line"># generate Dataset</span><br><span class="line">dds &lt;- DESeqDataSetFromMatrix(countData, colData, design=~Treatment, tidy=TRUE)</span><br><span class="line"></span><br><span class="line"># normlize using rlog mathod</span><br><span class="line">norm &lt;- rlog(dds,blind=FALSE)</span><br><span class="line">norm_matrix &lt;- assay(norm)</span><br><span class="line"></span><br><span class="line"># diff-exp analysis using DESeq2</span><br><span class="line">deg &lt;- DESeq(dds)</span><br><span class="line">res &lt;- results(deg,tidy=TRUE)</span><br></pre></td></tr></table></figure>
<h4 id="Heatmap-for-count-matrix"><a href="#Heatmap-for-count-matrix" class="headerlink" title="Heatmap for count matrix"></a>Heatmap for count matrix</h4><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">library(&quot;pheatmap&quot;)</span><br><span class="line">select &lt;- order(rowMeans(counts(deg,normalized=TRUE)), decreasing=TRUE)[1:25]</span><br><span class="line">df &lt;- as.data.frame(colData(deg)[,c(&quot;Treatment&quot;,&quot;Sample&quot;)])</span><br><span class="line">pheatmap(assay(norm)[select,], cluster_rows=FALSE, show_rownames=FALSE, cluster_cols=TRUE, annotation_col=df)</span><br></pre></td></tr></table></figure>

      
    </div>
    
    
    

    

    <div>
    
    <div>
    
        <div style="text-align:center;color: #ccc;font-size:14px;">-----The ---- end ----<i class="fa fa-paw"></i>--- Thanks --- for --- Reading----</div>
    
</div>

    
    </div>

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/assignment/" rel="tag"><i class="fa fa-tag"></i> assignment</a>
          
            <a href="/tags/techniques/" rel="tag"><i class="fa fa-tag"></i> techniques</a>
          
            <a href="/tags/bioinformatics/" rel="tag"><i class="fa fa-tag"></i> bioinformatics</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2018/05/16/wechat/" rel="next" title="分享一个微信机器人">
                <i class="fa fa-chevron-left"></i> 分享一个微信机器人
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2018/05/20/line4/" rel="prev" title="Linear Regression Assignment 9">
                Linear Regression Assignment 9 <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          

  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            Table of Contents
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            Overview
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <img class="site-author-image" itemprop="image"
                src="/images/avatar.png"
                alt="James Chen" />
            
              <p class="site-author-name" itemprop="name">James Chen</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">58</span>
                  <span class="site-state-item-name">posts</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">14</span>
                  <span class="site-state-item-name">categories</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">44</span>
                  <span class="site-state-item-name">tags</span>
                </a>
              </div>
            

          </nav>

          
            <div class="feed-link motion-element">
              <a href="/atom.xml" rel="alternate">
                <i class="fa fa-rss"></i>
                RSS
              </a>
            </div>
          

          
            <div class="links-of-author motion-element">
                
                  <span class="links-of-author-item">
                    <a href="https://github.com/james20141606" target="_blank" title="GitHub">
                      
                        <i class="fa fa-fw fa-github"></i>GitHub</a>
                  </span>
                
                  <span class="links-of-author-item">
                    <a href="mailto:xp-chen14@mails.tsinghua.edu.cn" target="_blank" title="E-Mail">
                      
                        <i class="fa fa-fw fa-envelope"></i>E-Mail</a>
                  </span>
                
            </div>
          

          
          

          
          

          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#My-assignment-in-bioinformatics-course"><span class="nav-number">1.</span> <span class="nav-text">My assignment in bioinformatics course</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#Codes-using-different-packages-producing-differential-expression-files"><span class="nav-number">2.</span> <span class="nav-text">Codes using different packages producing differential expression files</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#some-plot-from-DESeq2"><span class="nav-number">3.</span> <span class="nav-text">some plot from DESeq2</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#edgeR"><span class="nav-number">3.1.</span> <span class="nav-text">edgeR</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Wilcox-Mann-Whitney-U-Test"><span class="nav-number">3.2.</span> <span class="nav-text">Wilcox/Mann-Whitney-U Test</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#1-normalize-the-reads-by-library-size-edgeR"><span class="nav-number">3.2.1.</span> <span class="nav-text">1. normalize the reads by library size (edgeR)</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#2-identify-differential-expressed-gene-using-wilcoxon-test"><span class="nav-number">3.2.2.</span> <span class="nav-text">2. identify differential expressed gene using wilcoxon.test()</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#exploring-results"><span class="nav-number">3.3.</span> <span class="nav-text">exploring results</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#Heatmap-for-count-matrix"><span class="nav-number">3.3.0.1.</span> <span class="nav-text">Heatmap for count matrix</span></a></li></ol></li></ol></li></ol></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; <span itemprop="copyrightYear">2018</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">James Chen</span>

  
</div>



  <span class="post-meta-divider">|</span>




<div class="theme-info">
  <div class="powered-by"></div>
  <span class="post-count">博客全站共114.3k字</span>
</div>

        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  
    <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>
  

  
  
    <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>
  

  
  
    <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>
  

  
  
    <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>
  



  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.4"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.4"></script>



  
  

  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.4"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.4"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.4"></script>



  


  




	





  





  












  





  

  
  <script src="https://cdn1.lncld.net/static/js/av-core-mini-0.6.4.js"></script>
  <script>AV.initialize("klOjl0RBA8qP5IKgIXkOszBr-gzGzoHsz", "rCaN5wX4mjiRkRMzP95g7XHz");</script>
  <script>
    function showTime(Counter) {
      var query = new AV.Query(Counter);
      var entries = [];
      var $visitors = $(".leancloud_visitors");

      $visitors.each(function () {
        entries.push( $(this).attr("id").trim() );
      });

      query.containedIn('url', entries);
      query.find()
        .done(function (results) {
          var COUNT_CONTAINER_REF = '.leancloud-visitors-count';

          if (results.length === 0) {
            $visitors.find(COUNT_CONTAINER_REF).text(0);
            return;
          }

          for (var i = 0; i < results.length; i++) {
            var item = results[i];
            var url = item.get('url');
            var time = item.get('time');
            var element = document.getElementById(url);

            $(element).find(COUNT_CONTAINER_REF).text(time);
          }
          for(var i = 0; i < entries.length; i++) {
            var url = entries[i];
            var element = document.getElementById(url);
            var countSpan = $(element).find(COUNT_CONTAINER_REF);
            if( countSpan.text() == '') {
              countSpan.text(0);
            }
          }
        })
        .fail(function (object, error) {
          console.log("Error: " + error.code + " " + error.message);
        });
    }

    function addCount(Counter) {
      var $visitors = $(".leancloud_visitors");
      var url = $visitors.attr('id').trim();
      var title = $visitors.attr('data-flag-title').trim();
      var query = new AV.Query(Counter);

      query.equalTo("url", url);
      query.find({
        success: function(results) {
          if (results.length > 0) {
            var counter = results[0];
            counter.fetchWhenSave(true);
            counter.increment("time");
            counter.save(null, {
              success: function(counter) {
                var $element = $(document.getElementById(url));
                $element.find('.leancloud-visitors-count').text(counter.get('time'));
              },
              error: function(counter, error) {
                console.log('Failed to save Visitor num, with error message: ' + error.message);
              }
            });
          } else {
            var newcounter = new Counter();
            /* Set ACL */
            var acl = new AV.ACL();
            acl.setPublicReadAccess(true);
            acl.setPublicWriteAccess(true);
            newcounter.setACL(acl);
            /* End Set ACL */
            newcounter.set("title", title);
            newcounter.set("url", url);
            newcounter.set("time", 1);
            newcounter.save(null, {
              success: function(newcounter) {
                var $element = $(document.getElementById(url));
                $element.find('.leancloud-visitors-count').text(newcounter.get('time'));
              },
              error: function(newcounter, error) {
                console.log('Failed to create');
              }
            });
          }
        },
        error: function(error) {
          console.log('Error:' + error.code + " " + error.message);
        }
      });
    }

    $(function() {
      var Counter = AV.Object.extend("Counter");
      if ($('.leancloud_visitors').length == 1) {
        addCount(Counter);
      } else if ($('.post-title-link').length > 1) {
        showTime(Counter);
      }
    });
  </script>



  

  
<script>
(function(){
    var bp = document.createElement('script');
    var curProtocol = window.location.protocol.split(':')[0];
    if (curProtocol === 'https') {
        bp.src = 'https://zz.bdstatic.com/linksubmit/push.js';        
    }
    else {
        bp.src = 'http://push.zhanzhang.baidu.com/push.js';
    }
    var s = document.getElementsByTagName("script")[0];
    s.parentNode.insertBefore(bp, s);
})();
</script>


  
  

  
  


  

  

  
</body>
</html>
